docs: add styling section (#21525)
nuxt/nuxt: docs/1.getting-started/4.styling.md

---
navigation.icon: uil:palette
---

# Styling

Nuxt is highly flexible when it comes to styling. Write your own styles, or reference local and external stylesheet.
You can use CSS preprocessors, CSS frameworks, UI libraries, as well as Nuxt modules to style your application.

## Local Stylesheets

To add a local stylesheet, you can use the css property in the Nuxt configuration.
The natural place for your stylesheets is the `assets/` directory. You can then reference its path and Nuxt will include it to all the pages of your application.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  css: ['~/assets/css/main.css']
})
```

You can also reference stylesheets that are distributed through npm. Let's use the popular `animate.css` library as an example.

```bash
npm install animate.css
```

This package recommends that you use a javascript import statement to include it in your application. In Nuxt, you don't need to do that.
The package can just be referenced directly as a string in the css property of your Nuxt configuration.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  css: ['animate.css']
})
```

> ### Importing within components

Reviewer: ```suggestion
### Importing Within Components
```

You can import stylesheets in your pages and components directly, either using a javascript import or a css `@import` statement.
You can learn more about css imports in the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/CSS/@import).

```vue [pages/index.vue]
<script>
import '~/assets/css/first.css'
</script>
<style>
@import url("~/assets/css/second.css");
</style>
```

::alert{type=info}
Keep in mind that CSS imported like this will be injected globally. When the component is Loaded.
::

> ### Working with Fonts

Reviewer: ```suggestion
### Working With Fonts
```

Place your local fonts files in the `assets/` directory. You can then reference them in your stylesheets using `url()`.

```css [assets/css/main.css]
@font-face {
  font-family: 'FarAwayGalaxy';
  src: url('../fonts/FarAwayGalaxy.woff') format('woff');
  font-weight: normal;
  font-style: normal;
  font-display: swap;
}
```

Then reference your fonts by name in your stylesheets, pages or components:

```html
<style>
h1 {
  font-family: 'FarAwayGalaxy', sans-serif;
}
</style>
```

## External Stylesheets

You can include external stylesheets in your application by adding a link element in the head section of your nuxt.config file.
You can achieve this result using different methods.

You can manipulate the head with the `app.head` property of your Nuxt configuration:

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  app: {
    head: {
      link: [{ rel: "stylesheet", href: "https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css" }]
    }
}})
```

> ### Dynamically adding stylesheets

Reviewer: ```suggestion
### Dynamically Adding Stylesheets
```

You can use the useHead composable to dynamically set a value in your head in your code.

::ReadMore{link="/docs/api/composables/use-head"}
::

```ts
useHead({
  link: [{ rel: "stylesheet", href: "https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css" }]
})
```

> Nuxt uses `unhead` under the hood, therefore you can refer to it's full documentation [here](https://unhead.harlanzw.com/).

Reviewer: ```suggestion
Nuxt uses `unhead` under the hood, therefore you can refer to its full documentation [here](https://unhead.harlanzw.com/).
```

> ### Modifying the head with a Nitro Plugin

Reviewer: ```suggestion
### Modifying The Head With A Nitro Plugin
```

If you need more advanced control, you can intercept the renderered html with a hook and modify the head programatically.

> Create a plugin in `server/plugins/my-plugin.ts` like this :

Reviewer: ```suggestion
Create a plugin in `server/plugins/my-plugin.ts` like this:
```

```ts [server/plugins/my-plugin.ts]
export default defineNitroPlugin((nitro) => {
    nitro.hooks.hook("render:html", (html) => {
        html.head.push("<link rel ='stylesheet' href='https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css'>")
    })
})
```

## Using Preprocessors

> To use a preprocessor like SCSS, Sass, Less or Stylus, install it first.

Reviewer: ```suggestion
To use a preprocessor like SCSS, SASS, LESS or Stylus, install it first.
```

::code-group

```bash [SASS & SCSS]
npm install sass
```

```bash [LESS]
npm install less
```

```bash [Stylus]
npm install stylus
```

::

The natural place to write your stylesheets is the `assets` directory.

> You can then add your pre-processors source files to the `css` property of your Nuxt configuration.

Reviewer: ```suggestion
You can then add your preprocessors source files to the `css` property of your Nuxt configuration.
```

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  css: ['~/assets/scss/main.scss']
})
```

If you need to inject code in pre-processed files, like a [sass partial](https://sass-lang.com/documentation/at-rules/use#partials) with color variables, you can do so with the vite [preprocessors options](https://vitejs.dev/config/shared-options.html#css-preprocessoroptions).

> Create some partials in your `assets` directory :

Reviewer: ```suggestion
Create some partials in your `assets` directory:
```

::code-group

```scss [assets/_colors.scss]
$primary: #49240F;
$secondary: #E4A79D;
```

```sass [assets/_colors.sass]
$primary: #49240F
$secondary: #E4A79D
```

::

Then in your `nuxt.config` :

::code-group

```ts [SCSS]
export default defineNuxtConfig({
  vite: {
    css: {
      preprocessorOptions: {
        scss: {
          additionalData: '@use "@/assets/_colors.scss" as *;'
        }
      }
    }
  }
})
```

```ts [SASS]
export default defineNuxtConfig({
  vite: {
    css: {
      preprocessorOptions: {
        sass: {
          additionalData: '@use "@/assets/_colors.sass" as *\n'
        }
      }
    }
  }
})
```

::

> Nuxt uses Vite by default. If you wish to use webpack instead, refer to each pre-processor loader [documentation](https://webpack.js.org/loaders/sass-loader/).

Reviewer: ```suggestion
Nuxt uses Vite by default. If you wish to use Webpack instead, refer to each preprocessor loader [documentation](https://webpack.js.org/loaders/sass-loader/).
```

> By default Nuxt use Vite, if you wish to use webpack instead, refer to each pre-processor loader [documentation](https://webpack.js.org/loaders/sass-loader/).

Reviewer: ```suggestion
Nuxt uses Vite by default. If you wish to use webpack instead, refer to each pre-processor loader [documentation](https://webpack.js.org/loaders/sass-loader/).
```

## Single File Components (SFC) Styling

> One of the best thing about Vue and SFC is how great it is at naturally dealing with styling. You can directly write CSS or pre-processor code in the style block of your components file, therefore you will have fantastic developer experience without having to use something like CSS-in-JS. However if you wish to use CSS-in-JS, you can find 3rd party libraries and modules that support it, such as [pinceau](https://pinceau.dev/).

Reviewer: ```suggestion
One of the best thing about Vue and SFC is how great it is at naturally dealing with styling. You can directly write CSS or preprocessor code in the style block of your components file, therefore you will have fantastic developer experience without having to use something like CSS-in-JS. However if you wish to use CSS-in-JS, you can find 3rd party libraries and modules that support it, such as [pinceau](https://pinceau.dev/).
```

> One of the best thing about Vue and SFC is how great it is at naturally dealing with styling. You can directly write CSS or pre-processor code in the style block of your components file, therefore you will have fantastic developer experience without having to use something like CSS-in-JS. However if you wish to use CSS-in-JS, you can find 3rd party libraries and module that supports it, such as [pinceau](https://pinceau.dev/).

Reviewer: ```suggestion
One of the best thing about Vue and SFC is how great it is at naturally dealing with styling. You can directly write CSS or pre-processor code in the style block of your components file, therefore you will have fantastic developer experience without having to use something like CSS-in-JS. However if you wish to use CSS-in-JS, you can find 3rd party libraries and modules that support it, such as [pinceau](https://pinceau.dev/).
```

You can refer to the [Vue docs](https://vuejs.org/api/sfc-css-features.html) for a comprehensive reference about styling components in SFC.

> ### Class and Style bindings

Reviewer: ```suggestion
### Class And Style Bindings
```

You can leverage Vue SFC features to style your components with class and style attributes.

::code-group

```html [Ref and Reactive]
<script setup>
const isActive = ref(true)
const hasError = ref(false)
const classObject = reactive({
  active: true,
  'text-danger': false
})
</script>
<div class="static" :class="{ active: isActive, 'text-danger': hasError }"></div>
<div :class="classObject"></div>
```

```html [Computed]
<script setup>
const isActive = ref(true)
const error = ref(null)

const classObject = computed(() => ({
  active: isActive.value && !error.value,
  'text-danger': error.value && error.value.type === 'fatal'
}))
</script>
<div :class="classObject"></div>
```

```html [Array]
<script setup> 
const isActive = ref(true)
const errorClass = ref('text-danger')
</script>
<div :class="[{ active: isActive }, errorClass]"></div>
```

```html [Style]
<script setup>
const activeColor = ref('red')
const fontSize = ref(30)
const styleObject = reactive({ color: 'red', fontSize: '13px' })
</script>
<div :style="{ color: activeColor, fontSize: fontSize + 'px' }"></div>
<div :style="[baseStyles, overridingStyles]"></div>
<div :style="styleObject"></div>
```

::

> Refer to the [vue docs](https://vuejs.org/guide/essentials/class-and-style.html) for more information.

Reviewer: ```suggestion
Refer to the [Vue docs](https://vuejs.org/guide/essentials/class-and-style.html) for more information.
```

> ### Dynamic styles with v-bind

Reviewer: ```suggestion
### Dynamic Styles With `v-bind`
```

You can reference JavaScript variable and expression within your style blocks with the v-bind function.
The binding will be dynamic, meaning that if the variable value changes, the style will be updated.

```html
<template>
  <div class="text">hello</div>
</template>

<script setup>
const color = ref("red")
</script>

<style>
.text {
  color: v-bind(color);
}
</style>
```

> ### Scoped styles

Reviewer: ```suggestion
### Scoped Styles
```

> The scoped attribute allow you to style component in insolation. The styles declared with this attribute will only apply to this component.

Reviewer: ```suggestion
The scoped attribute allows you to style component in insolation. The styles declared with this attribute will only apply to this component.
```

```html
<style scoped>
.example {
  color: red;
}
</style>

<template>
  <div class="example">hi</div>
</template>
```

### CSS Modules

> You can use [css modules](https://github.com/css-modules/css-modules) with the module attribute.

Reviewer: ```suggestion
You can use [CSS Modules](https://github.com/css-modules/css-modules) with the module attribute. Access it with the injected `$style` variable.
```

```html
<template>
  <p :class="$style.red">This should be red</p>
</template>

<style module>
.red {
  color: red;
}
</style>
```

> ### Preprocessor support

Reviewer: ```suggestion
### Preprocessors Support
```

> SFC style blocks support pre-processors syntax. Vite come with built-in support for .scss, .sass, .less, .styl and .stylus files without configuration. You just need to install them first, and they will be available directly in SFC with the lang attribute.

Reviewer: ```suggestion
SFC style blocks support preprocessors syntax. Vite come with built-in support for .scss, .sass, .less, .styl and .stylus files without configuration. You just need to install them first, and they will be available directly in SFC with the lang attribute.
```

::code-group

```html [SCSS]
<style lang="scss">
  /* Write scss here */
</style>
```

```html [SASS]
<style lang="sass">
  /* Write sass here */
</style>
```

```html [LESS]
<style lang="less">
  /* Write less here */
</style>
```

> ```html [STYLUS]

Reviewer: ```suggestion
```html [Stylus]
```

<style lang="stylus">
  /* Write stylus here */
</style>
```

::

> You can refere to the [vite css docs](https://vitejs.dev/guide/features.html#css) and the [@vitejs/plugin-vue docs](https://github.com/vitejs/vite-plugin-vue/tree/main/packages/plugin-vue).

Reviewer: ```suggestion
You can refer to the [Vite CSS docs](https://vitejs.dev/guide/features.html#css) and the [@vitejs/plugin-vue docs](https://github.com/vitejs/vite-plugin-vue/tree/main/packages/plugin-vue).
```

> For webpack users, refer to the [vue loader docs](https://vue-loader.vuejs.org/).

Reviewer: ```suggestion
For Webpack users, refer to the [vue loader docs](https://vue-loader.vuejs.org/).
```

## Using PostCSS

Nuxt comes with postcss built-in. You can configure it in your `nuxt.config` file.

> ```ts [nxut.config.ts]

Reviewer: ```suggestion
```ts [nuxt.config.ts]
```

export default defineNuxtConfig({
    postcss: {
      plugins: {
        'postcss-nested': {}
        "postcss-custom-media": {}
      }
    }
})
```

For proper syntax highlighting in SFC, you can use the postcss lang attribute.

```html
<style lang="postcss">
  /* Write stylus here */
</style>
```

> By default, Nuxt comes with the following plugins already pre-configured :

Reviewer: ```suggestion
By default, Nuxt comes with the following plugins already pre-configured:
```

- [postcss-import](https://github.com/postcss/postcss-import): Improves the `@import` rule

> - [postcss-url](https://github.com/postcss/postcss-url) : Transform `url()` statements

Reviewer: ```suggestion
- [postcss-url](https://github.com/postcss/postcss-url): Transforms `url()` statements
```

> - [autoprefixer](https://github.com/postcss/autoprefixer) : Automatically add vendor prefixes

Reviewer: ```suggestion
- [autoprefixer](https://github.com/postcss/autoprefixer): Automatically adds vendor prefixes
```

> - [cssnano](https://cssnano.co/) : Minification and purge

Reviewer: ```suggestion
- [cssnano](https://cssnano.co/): Minification and purge
```

> ##  Leveraging Layouts for Multiple Styles

Reviewer: ```suggestion
##  Leveraging Layouts For Multiple Styles
```

If you need to style different parts of your application completely differently, you can use layouts.
Use different styles for different layouts.

```html
<template>
  <h1>Default Layout</h1>
  <slot />
</template>
<style>
.DefaultLayout {
  color: red;
}
</style>
```

::ReadMore{link="docs/guide/directory-structure/layouts"}
::

> ## Third party libraries and modules

Reviewer: ```suggestion
## Third Party Libraries And Modules
```

Nuxt isn't opinionated when it comes to styling and provides you with a wide variety of options. You can use any styling tool that you want, such as the popular [UnoCSS](https://unocss.dev/) or [Tailwind CSS](https://tailwindcss.com/).

The community and the Nuxt team have developped plenty of Nuxt modules to makes the integration easier.
Here's a non exhaustive list of modules that you can check-out to get started :

- [Uno CSS](https://nuxt.com/modules/unocss)
- [Tailwind CSS](https://nuxt.com/modules/tailwindcss)
- [Fontaine](https://github.com/nuxt-modules/fontaine)
- [Pinceau](https://pinceau.dev/)
- [NuxtLabs UI](https://nuxt.com/modules/nuxtlabs-ui)

> 

Reviewer: I think it would be great to emphasize more the Nuxt modules link with a filter preset for styling options

Author: Yes we could do that ! However I cherry picked this 5 because they are maintained by members of the Nuxt team if I'm not mistaken (danielroe, Atinux, antfu and Tahul).
However these 5 are not listed under the same section or even not listed, so I still think we should recommend them.

Reviewer: For sure 👍 Just adding a link would be enough like `[Explore more modules from the community](...)`

> Nuxt modules provide you with a good developer experience out of the box, but remember that if your favourite tool doesn't have a module, it doesn't mean that you can't use it with Nuxt! You can configure it yourself for your own project. Depending on the tool, you might need to use a [nuxt plugin](docs/guide/directory-structure/plugins), you might need to [make your own module](/docs/guide/going-further/modules). Share them with the [community](https://nuxt.com/modules) if you do!

Reviewer: ```suggestion
Nuxt modules provide you with a good developer experience out of the box, but remember that if your favourite tool doesn't have a module, it doesn't mean that you can't use it with Nuxt! You can configure it yourself for your own project. Depending on the tool, you might need to use a [Nuxt plugin](docs/guide/directory-structure/plugins) and/or [make your own module](/docs/guide/going-further/modules). Share them with the [community](https://nuxt.com/modules) if you do!
```

> ### Easily load webfonts

Reviewer: ```suggestion
### Easily Load Webfonts
```

You can use a 3rd party library such [UnoCSS](https://unocss.dev/integrations/nuxt) which comes with a [web fonts presets](https://unocss.dev/presets/web-fonts) to conveniently load fonts from common providers such as google fonts.

## Advanced

### Transitions

Nuxt comes with the same Transition element that Vue has, and also has support for the experimental [View Transitions API](docs/getting-started/transitions#view-transitions-api-experimental).

::ReadMore{link="docs/features/transitions"}
::

> ### Font advanced optimization

Reviewer: ```suggestion
### Font Advanced Optimization
```

If you need something more advanced, such as font metric fallbacks, you can create a Nuxt module to extend the build process or the Nuxt runtime.
You can have a look at the 3rd party module [Fontaine](https://github.com/nuxt-modules/fontaine) for an implementation example.

::alert{type="info"}

> Always remember to take advantage of the various tools and techniques available in the Web ecosystem at large to make styling your application easier and more efficient. Whether you're using native CSS, a pre-processor, postcss, a UI library or a module, Nuxt has got you covered. Happy styling!

Reviewer: ```suggestion
Always remember to take advantage of the various tools and techniques available in the Web ecosystem at large to make styling your application easier and more efficient. Whether you're using native CSS, a preprocessor, postcss, a UI library or a module, Nuxt has got you covered. Happy styling!
```

---
navigation.icon: uil:palette
---

# Styling

Nuxt is highly flexible when it comes to styling. Write your own styles, or reference local and external stylesheet.
You can use CSS preprocessors, CSS frameworks, UI libraries, as well as Nuxt modules to style your application.

## Local Stylesheets

To add a local stylesheet, you can use the css property in the Nuxt configuration.
The natural place for your stylesheets is the `assets/` directory. You can then reference its path and Nuxt will include it to all the pages of your application.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  css: ['~/assets/css/main.css']
})
```

You can also reference stylesheets that are distributed through npm. Let's use the popular `animate.css` library as an example.

```bash
npm install animate.css
```

This package recommends that you use a javascript import statement to include it in your application. In Nuxt, you don't need to do that.
The package can just be referenced directly as a string in the css property of your Nuxt configuration.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  css: ['animate.css']
})
```

> 

Reviewer 2: Shouldn't we recommend to use `app.vue` instead @danielroe ?

---
navigation.icon: uil:palette
---

# Styling

Nuxt is highly flexible when it comes to styling. Write your own styles, or reference local and external stylesheet.
You can use CSS preprocessors, CSS frameworks, UI libraries, as well as Nuxt modules to style your application.

## Local Stylesheets

If you're writing local stylesheets, the natural place to put them is the `assets/` directory.

### Importing Within Components

You can import stylesheets in your pages, layouts and components directly.
You can use a javascript import, or a css [`@import` statement](https://developer.mozilla.org/en-US/docs/Web/CSS/@import).

```vue [pages/index.vue]
<script>
import '~/assets/css/first.css'
</script>

> <style>

Reviewer 3: ```suggestion

<style>
```

@import url("~/assets/css/second.css");
</style>
```

::alert{type=info}
CSS imported like this will be inlined when the component is Loaded.
::

### The CSS Property

You can also use the css property in the Nuxt configuration.
The natural place for your stylesheets is the `assets/` directory. You can then reference its path and Nuxt will include it to all the pages of your application.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  css: ['~/assets/css/main.css']
})
```

Nuxt will create an `entry.css` file that will be referenced in the HTML pages served by Nuxt. A preload tag will be automatically injected.

::alert{type=warning}

> Due to the way browser works, importing in components will lead to better perfomances when it comes to [LCP](https://web.dev/lcp/).

Author: Should we say app.vue or layouts ?

::

You should use the CSS property if you explicitly want to include your style as an external stylesheet.

### Working With Fonts

> Place your local fonts files in the `assets/` directory. You can then reference them in your stylesheets using `url()`.

Author: In other places of the docs, we recommend placing them in assets. Should we consolidate to public/fonts everywhere ?

```css [assets/css/main.css]
@font-face {
  font-family: 'FarAwayGalaxy';
  src: url('../fonts/FarAwayGalaxy.woff') format('woff');
  font-weight: normal;
  font-style: normal;
  font-display: swap;
}
```

Then reference your fonts by name in your stylesheets, pages or components:

```vue
<style>
h1 {
  font-family: 'FarAwayGalaxy', sans-serif;
}
</style>
```

### Stylesheets Distributed Through NPM

You can also reference stylesheets that are distributed through npm. Let's use the popular `animate.css` library as an example.

```bash
npm install animate.css
```

Then you can reference it direclty in your pages, layouts and components:

```vue [pages/index.vue]
<script>
import 'animate.css'
</script>

> <style>

Reviewer 3: ```suggestion

<style>
```

@import url("animate.css");
</style>
```

The package can also be referenced as a string in the css property of your Nuxt configuration.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  css: ['animate.css']
})
```

## External Stylesheets

You can include external stylesheets in your application by adding a link element in the head section of your nuxt.config file.
You can achieve this result using different methods.

You can manipulate the head with the `app.head` property of your Nuxt configuration:

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  app: {
    head: {

>       link: [{ rel: "stylesheet", href: "https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css" }]

Reviewer 3: ```suggestion
      link: [{ rel: 'stylesheet', href: 'https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css' }]
```

::ReadMore{link="/docs/api/composables/use-head"}
::

```ts
useHead({
  link: [{ rel: "stylesheet", href: "https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css" }]
})
```

Nuxt uses `unhead` under the hood, therefore you can refer to its full documentation [here](https://unhead.harlanzw.com/).

### Modifying The Rendered Head With A Nitro Plugin

If you need more advanced control, you can intercept the renderered html with a hook and modify the head programatically.

Create a plugin in `server/plugins/my-plugin.ts` like this:

```ts [server/plugins/my-plugin.ts]
export default defineNitroPlugin((nitro) => {
    nitro.hooks.hook("render:html", (html) => {
        html.head.push("<link rel ='stylesheet' href='https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css'>")
    })
})
```

## Using Preprocessors

To use a preprocessor like SCSS, Sass, LESS or Stylus, install it first.

::code-group

```bash [Sass & SCSS]
npm install sass
```

```bash [LESS]
npm install less
```

```bash [Stylus]
npm install stylus
```

::

The natural place to write your stylesheets is the `assets` directory.

> You can then add your preprocessors source files to the `css` property of your Nuxt configuration.

Author: Should we recommend to import them in app.vue/layouts instead ?

> You can then add your preprocessors source files to the `css` property of your Nuxt configuration.

Reviewer 3: Probably. Though on second thought it feels to me more like we should handle this within Nuxt, which I'm pretty sure we can do.

> You can then add your preprocessors source files to the `css` property of your Nuxt configuration.

Author: This is not the most straightforward problem tbh.
Reading a little more into it, both approach (inline vs reference) have trade-offs, and the best thing for optimization would be to inline the critical css using a tool like https://github.com/addyosmani/critical. 

Personally I think we should make the css prop behaviour the same as importing, or deprecate it and have imports in app.vue/layouts as the default pattern.

If someone wants to have link tag with a href for their css, they should explicitly do that with  /public + head.

I'll change the recommendation for now.

For webpack users, refer to the [vue loader docs](https://vue-loader.vuejs.org/).

## Using PostCSS

Nuxt comes with postcss built-in. You can configure it in your `nuxt.config` file.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
    postcss: {
      plugins: {
        'postcss-nested': {}
        "postcss-custom-media": {}
      }
    }
})
```

For proper syntax highlighting in SFC, you can use the postcss lang attribute.

```vue
<style lang="postcss">
  /* Write stylus here */
</style>
```

By default, Nuxt comes with the following plugins already pre-configured:

- [postcss-import](https://github.com/postcss/postcss-import): Improves the `@import` rule
- [postcss-url](https://github.com/postcss/postcss-url): Transforms `url()` statements
- [autoprefixer](https://github.com/postcss/autoprefixer): Automatically adds vendor prefixes
- [cssnano](https://cssnano.co/): Minification and purge

##  Leveraging Layouts For Multiple Styles

If you need to style different parts of your application completely differently, you can use layouts.
Use different styles for different layouts.

```vue
<template>

>   <h1>Default Layout</h1>
>   <slot />

Reviewer 3: ```suggestion
  <div class="default-layout">
    <h1>Default Layout</h1>
    <slot />
  </div>
```

</template>

> <style>

Reviewer 3: ```suggestion

<style>
```

> .DefaultLayout {

Reviewer 3: ```suggestion
.default-layout {
```

The community and the Nuxt team have developed plenty of Nuxt modules to makes the integration easier.
You can discover them on the [modules section](https://nuxt.com/modules) of the website.
Here are a few modules that are maintained by members of the Nuxt team to help you get started:

- [Uno CSS](https://nuxt.com/modules/unocss) : Instant on-demand atomic CSS engine
- [Tailwind CSS](https://nuxt.com/modules/tailwindcss) : Utility-first CSS framework
- [Fontaine](https://github.com/nuxt-modules/fontaine) : Font metric fallback
- [Pinceau](https://pinceau.dev/) : Adaptable styling framework
- [NuxtLabs UI](https://nuxt.com/modules/nuxtlabs-ui)

Nuxt modules provide you with a good developer experience out of the box, but remember that if your favourite tool doesn't have a module, it doesn't mean that you can't use it with Nuxt! You can configure it yourself for your own project. Depending on the tool, you might need to use a [Nuxt plugin](docs/guide/directory-structure/plugins) and/or [make your own module](/docs/guide/going-further/modules). Share them with the [community](https://nuxt.com/modules) if you do!

### Easily Load Webfonts

> You can use a 3rd party library such [UnoCSS](https://unocss.dev/integrations/nuxt) which comes with a [web fonts presets](https://unocss.dev/presets/web-fonts) to conveniently load fonts from common providers such as google fonts.

Reviewer 3: ```suggestion
You can use [the Nuxt Google Fonts module](https://github.com/nuxt-modules/google-fonts) to load Google Fonts.

If you are using [UnoCSS](https://unocss.dev/integrations/nuxt), note that it comes with a [web fonts presets](https://unocss.dev/presets/web-fonts) to conveniently load fonts from common providers, including Google Fonts and more.
```

## Advanced

### Transitions

> Nuxt comes with the same Transition element that Vue has, and also has support for the experimental [View Transitions API](docs/getting-started/transitions#view-transitions-api-experimental).

Reviewer 3: ```suggestion
Nuxt comes with the same `<Transition>` element that Vue has, and also has support for the experimental [View Transitions API](docs/getting-started/transitions#view-transitions-api-experimental).
```

::ReadMore{link="docs/features/transitions"}
::

### Font Advanced Optimization

> If you need something more advanced, such as font metric fallbacks, you can create a Nuxt module to extend the build process or the Nuxt runtime.
> You can have a look at the 3rd party module [Fontaine](https://github.com/nuxt-modules/fontaine) for an implementation example.

Reviewer 3: ```suggestion
We would recommend using [Fontaine](https://github.com/nuxt-modules/fontaine) to reduce your [CLS](https://web.dev/cls/). If you need something more advanced, consider creating a Nuxt module to extend the build process or the Nuxt runtime.
```

::alert{type="info"}
Always remember to take advantage of the various tools and techniques available in the Web ecosystem at large to make styling your application easier and more efficient. Whether you're using native CSS, a preprocessor, postcss, a UI library or a module, Nuxt has got you covered. Happy styling!
::

### LCP Avanced optimizations

> If you're using the `css` property or external CSS files, you can do the following to speed-up the download of your global CSS files.

Reviewer 3: ```suggestion
You can do the following to speed-up the download of your global CSS files:
```

- Use a CDN so the files are physically closer to your users
- Compress your assets, ideally using Brotli
- Use HTTP2/HTTP3 for delivery
- Host your assets on the same domain (do not use a different subdomain)

Most of these things should be done for you automatically if you're using modern platforms like Cloudflare, Netlify or Vercel.
You can find an LCP optimization guide on [web.dev](https://web.dev/optimize-lcp/).

> If you're not using the `css` property, and your modules are not using it, you can completely remove the `entry.css` from the build process.

Reviewer 3: ```suggestion
If all of your CSS is inlined by Nuxt, you can (experimentally) completely stop external CSS files from being referenced in your rendered HTML.
```

---
navigation.icon: uil:palette
---

# Styling

Nuxt is highly flexible when it comes to styling. Write your own styles, or reference local and external stylesheets.
You can use CSS preprocessors, CSS frameworks, UI libraries and Nuxt modules to style your application.

## Local Stylesheets

If you're writing local stylesheets, the natural place to put them is the `assets/` directory.

### Importing Within Components

You can import stylesheets in your pages, layouts and components directly.
You can use a javascript import, or a css [`@import` statement](https://developer.mozilla.org/en-US/docs/Web/CSS/@import).

```vue [pages/index.vue]
<script>
import '~/assets/css/first.css'
</script>

<style>
@import url("~/assets/css/second.css");
</style>
```

::alert{type=info}
The stylesheets will be inlined in the HTML rendered by Nuxt.
::

### The CSS Property

> You can also use the css property in the Nuxt configuration.

Reviewer 2: ```suggestion
You can also use the `css` property in the Nuxt configuration.
```
